google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1alpha2/StreamWriter.java

```diff
@@ -0,0 +1,856 @@
+    messagesBatchLock = new ReentrantLock();
+    activeAlarm = new AtomicBoolean(false);
+    executor = builder.executorProvider.getExecutor();
+    backgroundResourceList = new ArrayList<>();
+    if (builder.executorProvider.shouldAutoClose()) {
+      backgroundResourceList.add(new ExecutorAsBackgroundResource(executor));
+    }
+    messagesWaiter = new Waiter(this.batchingSettings.getFlowControlSettings());
+    responseObserver = new AppendResponseObserver(this);
+
+    stubSettings =
+        BigQueryWriteSettings.newBuilder()
+            .setCredentialsProvider(builder.credentialsProvider)
+            .setExecutorProvider(builder.executorProvider)
+            .setTransportChannelProvider(builder.channelProvider)
+            .setEndpoint(builder.endpoint)
+            .build();
+    shutdown = new AtomicBoolean(false);
+    refreshAppend();
+  }
+
+  /** Stream name we are writing to. */
+  public String getStreamNameString() {
+    return streamName;
+  }
+
+  /**
+   * Schedules the writing of a message. The write of the message may occur immediately or be
+   * delayed based on the writer batching options.
+   *
+   * <p>Example of writing a message.
+   *
+   * <pre>{@code
+   * AppendRowsRequest message;
+   * ApiFuture<AppendRowsResponse> messageIdFuture = writer.append(message);
+   * ApiFutures.addCallback(messageIdFuture, new ApiFutureCallback<AppendRowsResponse>() {
+   *   public void onSuccess(AppendRowsResponse response) {
+   *     if (response.hasOffset()) {
+   *       System.out.println("written with offset: " + response.getOffset());
+   *     } else {
+   *       System.out.println("received an in stream error: " + response.error().toString());
+   *     }
+   *   }
+   *
+   *   public void onFailure(Throwable t) {
+   *     System.out.println("failed to write: " + t);
+   *   }
+   * }, MoreExecutors.directExecutor());
+   * }</pre>
+   *
+   * @param message the message in serialized format to write to BigQuery.
+   * @return the message ID wrapped in a future.
+   */
+  public ApiFuture<AppendRowsResponse> append(AppendRowsRequest message) {
+    Preconditions.checkState(!shutdown.get(), "Cannot append on a shut-down writer.");
+
+    final AppendRequestAndFutureResponse outstandingAppend =
+        new AppendRequestAndFutureResponse(message);
+    List<InflightBatch> batchesToSend;
+    messagesBatchLock.lock();
+    try {
+      batchesToSend = messagesBatch.add(outstandingAppend);
+      // Setup the next duration based delivery alarm if there are messages batched.
+      setupAlarm();
+      if (!batchesToSend.isEmpty()) {
+        for (final InflightBatch batch : batchesToSend) {
+          LOG.fine("Scheduling a batch for immediate sending.");
+          writeBatch(batch);
+        }
+      }
+    } finally {
+      messagesBatchLock.unlock();
+    }
+
+    return outstandingAppend.appendResult;
+  }
+
+  public void refreshAppend() throws IOException {
+    synchronized (this) {
```

Comment:
Isn't this a heavyweight operation?

Reply:
This is to re-connect. I don't think it is super heavy but yes, it costs more than normal in stream append.